src/meta/meta_ckpt.c

```diff
@@ -918,9 +922,16 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
         WT_ERR(__wt_metadata_update(session, WT_SYSTEM_CKPT_URI, buf->data));
     }
 
-    /* We also need to record the oldest timestamp in the metadata so we can set it on startup. */
+    /*
+     * We also need to record the oldest timestamp in the metadata so we can set it on startup. We
+     * should set the checkpoint's oldest timestamp as the minimum of the current oldest timestamp
+     * and the checkpoint timestamp.
+     *
+     * Cache the oldest timestamp to prevent us from reading two different values.
+     */
+    oldest_timestamp = txn_global->oldest_timestamp;
```

Comment:
I don't believe we need a read barrier here. The usage of oldest_timestamp happens after the definition so the compiler should not reorder that. `meta_ckpt_timestamp ` cannot change as we are still in a checkpoint.

Reply:
The compiler will very likely optimize out the local variable in this case. That's what I suggested the barrier would prevent.

Reply:
OK, that makes sense.

Reply:
@agorrod Though I put a read barrier here, I actually prefer to declare the variable as volatile as we are not trying to guarantee the order here but merely to prevent it from being optimized away.

In addition, this reminds me the `txnid` local variable in `__rec_upd_select` may suffer the same problem.

Reply:
We've oscillated on the `volatile` keyword, but moved away from using it, since the actual guarantees of volatile are not very useful. So, when we have a case like this we prefer to use barriers to enforce the expected behavior if possible.